### PR TITLE
Add new init methods

### PR DESCRIPTION
**Pull request by Kaixhin**
This is a (low priority) PR to improve the `init` functionality. Aims are:

- [x] Add a lookup function for various nonlinear gains.
- [x] Integrate the lookup function into `kaiming` init functions to make them less transparent/more general.
- [x] Add `eye` for preserving identity in Linear layers.
- [x] Add `dirac` for preserving identity in Convolutional layers (after discussion with @alykhantejani, this is less transparent than using `eye` in multiple ways as was done in `nninit`).
- [x] Make the docs more consistent/use MathJax/tidy references.
- [x] Add corresponding tests.

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -5,11 +5,39 @@
...
+        nonlinearity: the nonlinear function (`nn.functional` name)
+        param: optional parameter for the nonlinear function
+
+    Examples:
+        >>> gain = nn.init.gain('lrelu')
+    """
+    if gain == 'sigmoid':
+        return 1
+    elif gain == 'tanh':
+        return 5 / 3
```
Comment (fmassa) [inline torch/nn/init.py]: I think it might be better to use `5. / 3`, as Python 2.7 might not handle the casting to float

File: torch/nn/init.py [hunk 2]
```diff
@@ -5,11 +5,47 @@
...
 
+def calculate_gain(nonlinearity, param=None):
+    """Return the recommended gain value for the given nonlinearity function. The values are as follows:
+    ============ =========================================
+    nonlinearity gain
+    ============ =========================================
+    sigmoid      :math:`1`
+    tanh         :math:`5 / 3`
+    relu         :math:`\sqrt{2}`
+    lreaky_relu  :math:`\sqrt{2 / (1 + negative_slope^2)}`
```
Comment (Maratyszcza) [inline torch/nn/init.py]: Typo: `lreaky` -> `leaky`

File: torch/nn/init.py [hunk 3]
```diff
@@ -5,11 +5,47 @@
 from torch.autograd import Variable
 
 
+def calculate_gain(nonlinearity, param=None):
+    """Return the recommended gain value for the given nonlinearity function. The values are as follows:
+    ============ =========================================
+    nonlinearity gain
```
Comment (Maratyszcza) [inline torch/nn/init.py]: I'd suggest to add `nonlinearity == "linear"` (or `None`) for conv layers without activation

File: torch/nn/init.py [hunk 4]
```diff
@@ -55,14 +97,69 @@ def constant(tensor, val):
...
+        dirac(tensor.data, scaled=scaled)
+        return tensor
+
+    tensor.zero_()
+    val = 1.0 / tensor.size(1) if scaled else 1
+
+    if dimensions == 3:  # Temporal convolution
+        tensor[:, :, tensor.size(2) // 2].fill_(val)
+    elif dimensions == 4:  # Spatial convolution
+        tensor[:, :, tensor.size(2) // 2, tensor.size(3) // 2].fill_(val)
```
Comment (fmassa) [inline torch/nn/init.py]: I'm a bit confused about the definition of `dirac` in here. This filter is actually performing the sum over the input channels, so it doesn't give you the identity, but a transformation which preserves the sum over the channels.
The way I see it, in order to preserve the identity you'd need to consider individual feature planes, and not sum over them (something similar to grouped convolutions with number of groups = number of input channels).

**Comment by fmassa** [comment]
Looks good to me! Thanks Kai!

**Comment by Maratyszcza** [comment]
Thank you for this important work, Kai! Some extra suggestions:

There is some discrepancy in parameters which is worth unifying:
- `kaiming_uniform`/`kaiming_normal` accept parameter `a` (negative slope of ReLU)
- Other activation-dependent init functions accept parameter `gain`.

I'd prefer to have `gain` parameter for Kaiming inits as well, because with current `kaiming_*` interface is not compatible with non-ReLU gains (e.g. for linear activation).

Another idea: [Lasagne accepts](https://github.com/Lasagne/Lasagne/blob/master/lasagne/init.py#L210) `gain="relu"` as a shortcut for `math.sqrt(2)` gain, which is nice to avoid importing `math` module.

**Comment by Kaixhin** [comment]
I agree with unifying the `kaiming` init methods with others that use `gain` (this is how `nninit` was designed from the start), but the problem is that this is a breaking change. It's a pretty small one, but one that the core devs should weigh in on. I would say that any confusion about `gain` can be alleviated by proper documentation, so that itself shouldn't be an issue.

**Review by soumith — APPROVED** [review]

**Pull request merged**